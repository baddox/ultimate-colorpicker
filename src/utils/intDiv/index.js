import times from "lodash/times";
import round from "lodash/round";

const intDiv = (numerator, denomenator) => {
  const quotient = Math.trunc(numerator / denomenator);
  const extra = numerator % denomenator;
  const parts = times(denomenator, () => quotient);
  const extras = intDivRaw(denomenator, extra);
  return parts.map((quotient, index) => quotient + extras[index]);
};

export default intDiv;

export const intDivRaw = (count, extra) => {
  const extras = times(count, () => 0);
  if (extra > 0) {
    // The number of parts that will not need any extra added to them.
    const totalGap = count - extra;
    // The number of gaps to distribute the totalGap among.
    const gaps = totalGap < extra ? extra - 1 : extra + 1;
    // The width of each gap (likely not an integer).
    const gapWidth = totalGap / gaps;

    const offset = totalGap < extra ? 0 : gapWidth;

    // console.log("--------------------------------------");
    // console.log({
    //   count,
    //   extra,
    //   totalGap,
    //   gaps,
    //   gapWidth: round(gapWidth, 2),
    // });

    times(extra, extraIndex => {
      const extraIndexOffset = extraIndex + offset;
      const exactIndex = offset + gapWidth * extraIndex + extraIndex;
      const index = Math.round(exactIndex);
      // console.log({
      //   extraIndex,
      //   extraIndexOffset,
      //   exactIndex: round(exactIndex, 2),
      //   index,
      // });
      extras[index] += 1;
    });
  }
  return extras;
};
