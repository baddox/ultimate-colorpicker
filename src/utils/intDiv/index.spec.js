import intDiv, { intDivRaw } from "./";
import getSum from "lodash/sum";
import range from "lodash/range";

const toS = extras => {
  return extras.map(i => (i === 0 ? "_" : "-")).join("");
};

describe.only("intDiv", () => {
  // const total = 10;
  // const count = 3;
  // const parts = intDiv(total, count);
  // const sum = getSum(parts);

  // test("sum should equal total", () => {
  //   expect(sum).toEqual(total);
  // });

  // test("parts should be correct", () => {
  //   const expected = [3, 4, 3];
  //   expect(parts).toEqual(expected);
  // });

  const CASES = [
    [4, 0, `____`],
    [4, 1, `__-_`],
    [4, 2, `_--_`],
    [4, 3, `-_--`],

    [5, 0, `_____`],
    [5, 1, `__-__`],
    [5, 2, `_-_-_`],
    [5, 3, `-_-_-`],
    [5, 4, `--_--`],

    [9, 0, `_________`],
    [9, 1, `____-____`],
    [9, 2, `__-___-__`],
    [9, 3, `__-_-__-_`],
    [9, 4, `_-_-_-_-_`],
    [9, 5, `-_-_-_-_-`],
    [9, 6, `-_--_--_-`],
    [9, 7, `--_---_--`],
    [9, 8, `----_----`],
  ];

  describe("intDivRaw", () => {
    CASES.forEach(([count, extra, expected]) => {
      const label = `intDivRaw(${count}, ${extra})`;
      const extras = intDivRaw(count, extra);
      const string = toS(extras);

      test(`${label} parts should be correct`, () => {
        expect(string).toEqual(expected);
      });
    });
  });

  describe.skip("intDivRaw", () => {
    range(7, 8).forEach(count => {
      range(count, count * 2).forEach(total => {
        const label = `intDivRaw(${total}, ${count})`;
        const parts = intDivRaw(total, count);
        const sum = getSum(parts);

        test(`${label} sum should equal total`, () => {
          expect(sum).toEqual(total);
        });

        test(`${label} parts should be correct`, () => {
          const string = toS(parts);
          expect(`   ${string}   `).toMatchSnapshot();
        });

        // test("parts should be correct", () => {
        //   const expected = [3, 4, 3];
        //   expect(parts).toEqual(expected);
        // });
      });
    });
  });
});
