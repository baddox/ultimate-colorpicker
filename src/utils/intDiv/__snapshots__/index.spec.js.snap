// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`intDiv intDivRaw intDivRaw(7, 7) parts should be correct 1`] = `"   -------   "`;

exports[`intDiv intDivRaw intDivRaw(8, 7) parts should be correct 1`] = `"   ---_----   "`;

exports[`intDiv intDivRaw intDivRaw(9, 7) parts should be correct 1`] = `"   --_---_--   "`;

exports[`intDiv intDivRaw intDivRaw(10, 7) parts should be correct 1`] = `"   -_--_--_--   "`;

exports[`intDiv intDivRaw intDivRaw(11, 7) parts should be correct 1`] = `"   -_--_-_--_-   "`;

exports[`intDiv intDivRaw intDivRaw(12, 7) parts should be correct 1`] = `"   -_-_-_--_-_-   "`;

exports[`intDiv intDivRaw intDivRaw(13, 7) parts should be correct 1`] = `"   -_-_-_-_-_-_-   "`;
